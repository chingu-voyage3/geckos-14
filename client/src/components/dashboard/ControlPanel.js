/** ControlPanel Component:
ControlPanel component is a kind of toolbar of DevicePanel and VizPanel.
So far it only display a device/viz adding form
TODO: Move the form to specific Component
TODO: Add other controls like reorder/save

Testing Data:
  Device
    Name: test
    Type: LCD
    Url: http://devices.webofthings.io/pi/actuators/display/

  Viz
    Name: test
    Model: VictoryLine or VictoryBar
    Device: No need to fill
*/

import React, { Component } from 'react';
import Param from './Param';
import * as d from '../../tempData';

class ControlPanel extends Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.handleAdd = this.handleAdd.bind(this);
    this.state = {
      device: {
        name: '',
        type: '',
        url: ''
      },
      viz: {
        name: '',
        model: '',
        url: '',
        x: 'date',
        y: 'hum',
        data: d.testData,
        options: []
      }
    };
  }
  handleChange(event) {
    const value = event.target.value;
    const name = event.target.name;
    // console.log(event.target.name);
    // eslint-disable-next-line
    this.props.parent === 'device'
      ? this.setState({ device: { ...this.state.device, [name]: value } })
      : this.setState({ viz: { ...this.state.viz, [name]: value } });
  }

  handleAdd() {
    if (this.props.parent === 'device') {
      this.props.actions(this.state.device);
      this.setState({
        device: {
          name: '',
          type: '',
          url: ''
        }
      });
    } else {
      this.props.actions(this.state.viz);
      this.setState({
        viz: {
          name: '',
          model: '',
          data: d.testData,
          x: 'date',
          y: 'hum',
          url: ''
        }
      });
    }
  }
  renderParams(params) {
    return params.map(param => {
      return (
        <Param
          key={param.name}
          {...param}
          value={this.state[param.name]}
          onChange={this.handleChange}
        />
      );
    });
  }
  render() {
    // console.log('params', this.props.params);
    return (
      <div className="control-panel">
        <div className="add">
          {this.renderParams(this.props.params)}
          <button className="add-action" onClick={this.handleAdd}>
            Add
          </button>
        </div>
      </div>
    );
  }
}

export default ControlPanel;
