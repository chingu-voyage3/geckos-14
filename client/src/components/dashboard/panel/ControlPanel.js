/** ControlPanel Component:
ControlPanel component is a kind of toolbar of ThingPanel
So far it only display a thing adding form.
State is used to allow click detection and to fill the Params Form for Edit or Deletion
TODO: Add other controls like reorder/save
Testing Data:
  Thing
    Name: test
    Type: LCD
    Url: http://things.webofthings.io/pi/actuators/display/
*/

import React, { Component } from 'react';
import Param from './Param';
import Action from './Action';
import * as d from '../../../tempData.js';

class ControlPanel extends Component {
  constructor(props) {
    super(props);
    // state initialisation for controlled component
    this.state = {
      thing: {
        id: '',
        name: ''
      },
      thingSelected: false
    };
  }
  // shouldComponentUpdate(nextProps, nextState) {
  //   return this.props.selected !== nextProps.selected;
  // }
  componentWillReceiveProps(nextProps) {
    // console.log('Thing nextProps', nextProps);
    if (nextProps.selected.parent === 'thing') {
      // check if selected Item and Item is Thing;
      if (this.state.thing.name !== nextProps.selected.item.name) {
        // console.log('different');
        // Check initial state
        if (!this.state.thingSelected) {
          // console.log('Thing - move to true');
          this.setState({
            thing: nextProps.selected.item,
            thingSelected: true
          });
        } else {
          // if is selected and id is same then we unselect and clear Params form
          // console.log('Thing - move to false');
          this.clearState();
        }
      }
    }
  }
  // allowing controlled components by catching all change events and updating state
  handleChange = event => {
    const value = event.target.value;
    const name = event.target.name;
    // eslint-disable-next-line
    this.setState({ thing: { ...this.state.thing, [name]: value } });
  };
  // Handles discovery
  handleDiscover = () => {
    this.props.actions.discover(this.state.thing.id);
    // this.setState({
    //   selectedThing: true
    // });
  };
  // Handles delering
  handleDel = () => {
    // console.log('Del Item');
    this.props.actions.delThing(this.state.thing.id);
    this.clearState();
  };
  // clearing State
  clearState = () => {
    // console.log('Clearing State');
    this.setState({
      thing: {
        id: '',
        name: ''
      },
      thingSelected: false
    });
  };
  // helper function rendering Params
  renderParams = params => {
    return params.map(param => {
      return (
        <Param
          key={param.name}
          {...param}
          value={this.state.thing[param.name]}
          onChange={this.handleChange}
          options={param.options}
        />
      );
    });
  };
  // helper function rendering Actions
  renderActions = (actions, onclicks) => {
    return actions.map(action => {
      if (action.onSelect) {
        return (
          <div key={action.name} className={this.state.thingSelected ? 'visible' : 'hidden'}>
            <Action {...action} onclick={onclicks[action.name]} />
          </div>
        );
      }
      return (
        <div key={action.name} className={this.state.thingSelected ? 'hidden' : 'visible'}>
          <Action {...action} onclick={onclicks[action.name]} />
        </div>
      );
    });
  };

  render() {
    const onclicks = { Discover: this.handleDiscover, Edit: this.handleEdit, Del: this.handleDel };
    return (
      <div className="control-panel">
        <div className="control-panel-actions">
          {this.renderActions(d.devPanelActions, onclicks)}
        </div>
        <div className="control-panel-params">{this.renderParams(this.props.params)}</div>
      </div>
    );
  }
}

export default ControlPanel;
