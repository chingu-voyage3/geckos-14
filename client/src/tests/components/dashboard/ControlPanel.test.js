import React from 'react';
import ControlPanel from '../../../components/dashboard/ControlPanel';

import { shallow } from 'enzyme';
const params = [
  {
    label: 'Name :',
    name: 'name'
  },
  {
    label: 'Type :',
    name: 'type'
  },
  {
    label: 'Url :',
    name: 'url'
  }
];
const addViz = viz => {
  // eslint-disable-next-line no-console
  console.log('viz to be Added', viz);
};

it('renders without crashing', () => {
  const wrapper = shallow(<ControlPanel params={params} actions={addViz} parent="device" />);
  expect(toJson(wrapper)).toMatchSnapshot();
});
