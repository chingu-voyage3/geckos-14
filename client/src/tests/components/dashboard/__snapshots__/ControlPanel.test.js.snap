// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="control-panel"
>
  <div
    className="add"
  >
    <Param
      key="name"
      label="Name :"
      name="name"
      onChange={[Function]}
    />
    <Param
      key="type"
      label="Type :"
      name="type"
      onChange={[Function]}
    />
    <Param
      key="url"
      label="Url :"
      name="url"
      onChange={[Function]}
    />
    <button
      className="add-action"
      onClick={[Function]}
    >
      Add
    </button>
  </div>
</div>
`;
